File: backend/test/c/test_constants.c
#include <assert.h>

#include "test_constants_c.h"

int main()
{
  {
    enum test_simple_enum e;

    e = TEST_ENUM_1;
    assert(e == 1u);

    e = TEST_ENUM_2;
    assert(e == 2u);
  }

  {
    enum test_scoped_enum e;

    e = TEST_SCOPED_ENUM_SCOPED_ENUM_1;
    assert(e == 1u);

    e = TEST_SCOPED_ENUM_SCOPED_ENUM_2;
    assert(e == 2u);
  }

  assert(TEST_ANONYMOUS_ENUM_1 == 1u);
  assert(TEST_ANONYMOUS_ENUM_2 == 2u);

  assert(TEST_DUPLICATE_ENUM_1 == 1u);
  assert(TEST_DUPLICATE_ENUM_2 == 1u);

  assert(TEST_ENUM_IN_ANONYMOUS_NAMESPACE_1 == 1u);
  assert(TEST_ENUM_IN_ANONYMOUS_NAMESPACE_2 == 2u);

  assert(get_macro_const() == 1);
  assert(get_macro_expr() == ~(1 << 3));

  return 0;
}

File: backend/test/rust/test_constants.rs
use std::os::raw::*;

include!("test_constants_rust.rs");

fn main() {
    {
        let e = TestSimpleEnum::Enum1 as u32;
        assert!(e == 1);

        let e = TestSimpleEnum::Enum2 as u32;
        assert!(e == 2);

        let e = TestSimpleEnum::Enum1;

        let e_str = match e {
          TestSimpleEnum::Enum1 => "Enum1",
          TestSimpleEnum::Enum2 => "Enum2",
        };

        assert!(e_str == "Enum1");
    }

    {
        let e = TestScopedEnum::ScopedEnum1 as u32;
        assert!(e == 1);

        let e = TestScopedEnum::ScopedEnum2 as u32;
        assert!(e == 2);
    }

    assert!(TEST_ANONYMOUS_ENUM_1 == 1);
    assert!(TEST_ANONYMOUS_ENUM_2 == 2);

    assert!(TEST_DUPLICATE_ENUM_1 == 1);
    assert!(TEST_DUPLICATE_ENUM_2 == 1);

    {
        let e = TestEnumInAnonymousNamespace::EnumInAnonymousNamespace1 as u32;
        assert!(e == 1);

        let e = TestEnumInAnonymousNamespace::EnumInAnonymousNamespace2 as u32;
        assert!(e == 2);
    }

    // XXX macros
}
